src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
Please review the implementation as this IS the exact same scenario that we have. They have a way to inject the mandatory (scoped) ControllerContext into the controller, regardless of the controller activator in use, and whether there are accessible constructors or not, as this is an extremely important *facet* of the controller.
If you still think it's not the same scenario, I'd like to understand why, because I've been reviewing it thoroughly for months (since I first created the scoped activation context PR), and the more I look at it, the more I believe is the exact same scenario.
BTW, I'm not suggesting to expose the IGrainPropertyActivator abstraction (in case you are worried they might replace it and make our system grains not work), just make the DefaultGrainPropertyActivator the only non-replaceable implementation for now. We can guarantee that the pattern is respected without any gray areas.

Moving this to property injection is not just moving the problem from one place to another. Is avoiding it entirely. No matter which container they choose, we can always know which property we are injecting one at a time, and read the metadata at the time we are building the facet and injecting it. There is no guessing which member is being constructed, nor limiting how grains are resolved.

> Given a choice between providing clean, intuitive, and maintainable capabilities to our users or arbitrarily confining our solution to existing common patterns, I will choose a better user experience every time.

I guarantee they have the same driving function... it's not because they are fans of less-popular practices that they chose this approach. Also, as I mentioned, in our case there is no additional temporal coupling which is the main reason property injection is less favorable. This approach is also not less intuitive (and far more documented given how MVC is). With regards to the maintainable or confining solution comment, the proven approach wins that by an order of magnitude, I thought there was no question there. What remains is (subjectively) preferring constructor injection regardless of the reasons (still haven't heard a reason why constructor injection would be better for this case, other than personal preference).